pkg/finality-grandpa/voter.go

```diff
@@ -0,0 +1,1125 @@
+				// otherwise the base could be too eagerly set forward.
+				if state.Finalized != nil {
+					fNum := state.Finalized.Number
+					if fNum > v.lastFinalizedInRounds.Number {
+						v.lastFinalizedInRounds = *state.Finalized
+					}
+				}
+
+				err := v.env.Completed(
+					justCompleted.roundNumber(),
+					justCompleted.roundState(),
+					justCompleted.dagBase(),
+					justCompleted.historicalVotes(),
+				)
+				if err != nil {
+					v.inner.Unlock()
+					return err
+				}
+
+				v.inner.pastRounds.Push(v.env, justCompleted)
+
+				oldBest := v.inner.bestRound
+				v.inner.bestRound = newBest
+				v.inner.pastRounds.Push(v.env, oldBest)
+
+				if processCatchUpOutcome != nil {
+					processCatchUpOutcome(newCatchUpProcessingOutcome(CatchUpProcessingOutcomeGood{}))
+				}
+				v.inner.Unlock()
+			}
+		default:
+			break loop
+		}
+	}
+	return nil
+}
+
+// process the logic of the best round.
+func (v *Voter[Hash, Number, Signature, ID]) processBestRound(waker *waker) (bool, error) {
+	// If the current `bestRound` is completable and we've already precommitted,
+	// we start a new round at `bestRound + 1`.
+	{
+		v.inner.Lock()
+
+		var shouldStartNext bool
+		completable, err := v.inner.bestRound.poll(waker)
+		if err != nil {
+			return true, err
+		}
+
+		var precomitted bool
+		state := v.inner.bestRound.State()
+		if state != nil {
+			_, precomitted = v.inner.bestRound.State().(statePrecommitted)
+		}
+
+		shouldStartNext = completable && precomitted
+
+		if !shouldStartNext {
+			v.inner.Unlock()
+			return false, nil
+		}
+
+		log.Tracef("Best round at %v has become completable. Starting new best round at %v",
+			v.inner.bestRound.roundNumber(),
+			v.inner.bestRound.roundNumber()+1,
+		)
+		v.inner.Unlock()
+	}
+
+	err := v.completedBestRound()
+	if err != nil {
+		return true, err
+	}
+
+	// round has been updated. so we need to re-poll.
+	return v.poll(waker)
+}
+
+func (v *Voter[Hash, Number, Signature, ID]) completedBestRound() error {
+	v.inner.Lock()
+	defer v.inner.Unlock()
+
+	err := v.env.Completed(
+		v.inner.bestRound.roundNumber(),
+		v.inner.bestRound.roundState(),
+		v.inner.bestRound.dagBase(),
+		v.inner.bestRound.historicalVotes(),
+	)
+	if err != nil {
+		return err
+	}
+
+	oldRoundNumber := v.inner.bestRound.roundNumber()
+
+	nextRound := newVotingRound(
+		oldRoundNumber+1,
+		v.voters,
+		v.lastFinalizedInRounds,
+		v.inner.bestRound.bridgeState(),
+		v.inner.bestRound.FinalizedSender(),
+		v.env,
+	)
+
+	oldBest := v.inner.bestRound
+	v.inner.bestRound = nextRound
+	v.inner.pastRounds.Push(v.env, oldBest)
+	return nil
+}
+
+func (v *Voter[Hash, Number, Signature, ID]) setLastFinalizedNumber(finalizedNumber Number) bool {
+	if finalizedNumber > v.lastFinalizedNumber {
+		v.lastFinalizedNumber = finalizedNumber
+		return true
+	}
+	return false
+}
+
+func (v *Voter[Hash, Number, Signature, ID]) Start() error { //skipcq: RVV-B0001
+	v.wg.Add(1)
+	defer v.wg.Done()
+	waker := newWaker()
```

Comment:
You are using this `Start()` as a goroutine. You can add a goroutine to a waitgroup inside that same goroutine.
do wg.Add(1) outside.
pass wg to start (`Start(wg)`)
and them `defer wg.Done` inside it

https://blog.logrocket.com/concurrency-patterns-golang-waitgroups-goroutines/

Reply:
The function is intended to be synchronous.  The caller of this function needs to explicitly call `Start` in a goroutine if they want to run some other code in parallel.  Not sure why I need to pass in a wait group to this, if the intention is the call should be synchronous.

Reply:
Again the intention of this `Start()` method is that it is supposed to be synchronous, in that it should run until `Stop()` is called or it fails during execution and returns an appropriate error.  Maybe I should rename it to `Run` instead? The intention of the internal waitgroup is that `Start()` could still be polling when `Stop()` is called, and we are waiting for the goroutines to complete before exiting the `Start` function.   The waitgroup is only `Wait()`ed on in `Stop()`.